File: source/lexer.cpp
#include <iostream>
#include <sstream>
#include <stack>
#include "lexer.h"
using namespace Lexer;


Node * Node::make(Type type, const std::string & text, const Range & range)
{
	//Allocate a new node accordingly.
	Node * node = NULL;
	switch (type) {
		case kSymbol: {
			
			//Check whether this symbol starts a new group.
			Group::Subtype gst = Group::subtypeForOpeningSymbol(text);
			if (gst != Group::kInvalidGroup)
				node = new Group(gst);
		} break;
	}
	if (!node && type != kInvalid)
		node = new Node(type);
	
	//Apply the basic attributes to the node.
	if (node) {
		node->text = text;
		node->range = range;
	}
	return node;
}

Node::Node(Type t) : type(t)
{
	next = NULL;
}

Node::~Node()
{
	if (next) {
		delete next;
		next = NULL;
	}
}

bool Node::isIdentifier(const std::string & name) const
{
	return (type == kIdentifier && (name.empty() || text == name));
}

bool Node::isSymbol(const std::string & name) const
{
	return (type == kSymbol && (name.empty() || text == name));
}

bool Node::isGroup() const
{
	return (type == kGroup);
}

std::string Node::describe(int depth, int indent)
{
	return text;
}


Group::Group(Subtype s) : Node(kGroup), subtype(s)
{
	firstChild = NULL;
	lastChild = NULL;
}

Group::~Group()
{
	if (firstChild) {
		delete firstChild;
		firstChild = NULL;
	}
	lastChild = NULL;
}

void Group::addChild(Node * c)
{
	if (!c) return;
	if (lastChild)
		lastChild->next = c;
	lastChild = c;
	if (!firstChild)
		firstChild = lastChild;
}

const std::string Group::getOpeningSymbol()
{
	switch (subtype) {
		case kBrackets: return "["; break;
		case kBraces: return "{"; break;
		case kParanthesis: return "("; break;
	}
	return "";
}

const std::string Group::getClosingSymbol()
{
	switch (subtype) {
		case kBrackets: return "]"; break;
		case kBraces: return "}"; break;
		case kParanthesis: return ")"; break;
	}
	return "";
}

Group::Subtype Group::subtypeForOpeningSymbol(const std::string & os)
{
	if (os == "[") return kBrackets;
	if (os == "{") return kBraces;
	if (os == "(") return kParanthesis;
	return kInvalidGroup;
}

std::string Group::describe(int depth, int indent)
{
	std::string brk = "\n";
	brk += std::string(indent * 4, ' ');
	std::stringstream s;
	s << getOpeningSymbol();
	if (subtype == kBraces && depth != 0)
		s << brk;
	if (depth != 0) {
		Node * c = firstChild;
		bool first = true;
		while (c) {
			if (!first)
				s << " ";
			first = false;
			s << c->describe((depth > 0 ? depth - 1 : depth), indent + 1);
			if (c->type == Node::kSymbol && c->text == ";") {
				s << brk;
				first = true;
			}
			c = c->next;
		}
	} else {
		s << " ... ";
	}
	s << getClosingSymbol();
	return s.str();
}


Identifier::Identifier() : Node(kIdentifier) {}


//These are ranges depicting symbols.
const static struct { int s; int e; } symbolRanges[] = {
	{0x21, 0x2F},
	{0x3A, 0x3F},
	{0x5B, 0x5E},
	{0x7B, 0x7E},
	{0, 0}
};

//These are valid combinations of multiple symbols.
const static char * symbolCombinations[] = {
	"=>",
	NULL
};


/** Parses the given input stream and returns the root group node containing the
 * parsed nodes. */
Group * Lexer::parse(std::istream * input)
{
	//Allocate the root group node.
	Group * root = new Group(Group::kRoot);
	
	//Create the group stack so we can have nested nodes.
	std::stack<Group *> groupStack;
	groupStack.push(root);
	
	//We need to keep track of the context we're currently in, since each
	//requries some different form of processing.
	enum Context {
		kCodeContext,
		kSingleLineCommentContext,
		kMultiLineCommentContext,
		kStringContext,
	} context = kCodeContext;
	
	//We need to keep a string around so we can buffer up the characters and
	//keep track of what we are expecting to read.
	Node::Type bufferType = Node::kInvalid;
	Range bufferRange;
	std::string buffer;
	
	//Keep track of the location within the file.
	Location loc;
	
	//Read the input character by character until nothing is left.
	int pc = 0, ppc = 0;
	while (!input->eof()) {
		int c = input->get();
		
		//Advance the location tracker.
		if (c == '\n') {
			loc.line++;
			loc.col = 0;
		}
		
		//Process the character based on the current context.
		Node::Type newBufferType = Node::kInvalid;
		bool wrapUpAnyway = false;
		switch (context) {
				
			//Inside normal code.
			case kCodeContext: {
				
				//Check for comments.
				if (c == '/' && input->peek() == '/') {
					context = kSingleLineCommentContext;
					break;
				}
				if (c == '/' && input->peek() == '*') {
					context = kMultiLineCommentContext;
					break;
				}
				
				//Check for strings.
				if (c == '"') {
					context = kStringContext;
					break;
				}
				
				//Check for symbols.
				bool isSymbol = false;
				for (int i = 0; symbolRanges[i].s != 0; i++) {
					if (c >= symbolRanges[i].s && c <= symbolRanges[i].e) {
						isSymbol = true;
						break;
					}
				}
				if (isSymbol) {
					//Check whether the character can be combined with what's
					//already in the buffer.
					newBufferType = Node::kSymbol;
					std::string lookaheadBuffer = buffer;
					lookaheadBuffer += c;
					wrapUpAnyway = true;
					for (int i = 0; symbolCombinations[i] != 0; i++) {
						if (lookaheadBuffer == symbolCombinations[i]) {
							wrapUpAnyway = false;
							break;
						}
					}
					break;
				}
				
				//Check for numbers.
				if (c >= 0x30 && c <= 0x39) {
					newBufferType =  Node::kNumeric;
					break;
				}
				
				//Check for an identifier, which basically is anything except
				//the above and whitespaces.
				if (c > 0 && c != '\n' && c != '\t' && c != ' ' && c != '\r') {
					newBufferType = Node::kIdentifier;
					break;
				}
			} break;
				
			//Inside comments.
			case kSingleLineCommentContext: {
				if (c == '\n')
					context = kCodeContext;
			} break;
				
			case kMultiLineCommentContext: {
				if (c == '/' && pc == '*')
					context = kCodeContext;
			} break;
			
			//Inside strings.
			case kStringContext: {
				//Check whether we've reached the end of the string in which
				//case we switch back to code context, leaving the kind invalid
				//which will cause the token to be wrapped up.
				if (c == '"' && pc != '\\' && ppc != '\\')
					context = kCodeContext;
				
				//Otherwise we mark this character as part of a string token.
				else if (c > 0)
					newBufferType = Node::kString;
			} break;
		}
		
		//Rotate the character archive.
		ppc = pc;
		pc = c;
		
		//Numeric tokens are allowed in an identifier buffer.
		if (bufferType == Node::kIdentifier &&
			newBufferType == Node::kNumeric)
			newBufferType = bufferType;
		
		//Numeric buffers suck up all identifier and some symbol tokens, since
		//the value's validity is checked at a later stage.
		if (bufferType == Node::kNumeric) {
			if (newBufferType == Node::kIdentifier || c == '.') {
				newBufferType = bufferType;
				wrapUpAnyway = false;
			}
		}
		
		//If the buffer type and this token tpye disagree, wrap up the current
		//buffer.
		if ((bufferType != newBufferType || wrapUpAnyway) && !buffer.empty()) {
			
			//If the current token is a symbol and it closes the topmost group,
			//remove the group from the stack and ignore the symbol.
			if (bufferType == Node::kSymbol &&
				groupStack.top()->getClosingSymbol() == buffer) {
				groupStack.pop();
			} else {
				
				//Create a new node with the buffer content.
				Node * n = Node::make(bufferType, buffer, bufferRange);
				if (n) {
					
					//Add the node to the topmost group.
					groupStack.top()->addChild(n);
					
					//If the returned node is a group node, add it to the top of the
					//group stack so its subnodes get added to it.
					if (n->type == Node::kGroup)
						groupStack.push((Group *)n);
				}
			}
				
			//Clear the buffer and buffer kind.
			buffer.clear();
			bufferType = Node::kInvalid;
		}
		
		//If the current token is valid, set it as the current buffer kind and
		//move the character into the buffer.
		if (newBufferType != Node::kInvalid) {
			if (bufferType == Node::kInvalid)
				bufferRange.start = loc;
			bufferType = newBufferType;
			buffer += c;
		}
		
		//Increase the column counter.
		loc.col++;
		loc.offset++;
		
		//Adjust the buffer range's end now, so that the character we've read is
		//also counted in.
		bufferRange.end = loc;
	}
	
	//Return the root group.
	return root;
}

File: source/Lexer.php
<?php

class Lexer
{
	public $file;
	public $tokens;
	public $flatTokens;
	public $issues;
	
	public function run()
	{
		assert($this->file);
		$input = $this->file->content;
		
		$this->tokens = array();
		$this->flatTokens = array();
		$this->issues = array();
		
		//Accumulation buffer for tokens.
		$bufferType  = null;
		$bufferRange = new Range;
		$bufferRange->source = $this->file;
		$buffer      = "";
		
		//Location tracking.
		$loc = new Location;
		
		//Keep a group stack.
		$groupStack = array();
		
		//Iterate through every character in the input string.
		$context = 'code';
		$len = strlen($input);
		$pc = $c = 0;
		for ($i = 0; $i < $len + 1; $i++) {
			$ppc = $pc;
			$pc  = $c;
			$c   = ($i   < $len ? $input[$i]   : null);
			$nc  = ($i+1 < $len ? $input[$i+1] : null);
			
			//Process the character based on the context.
			$newBufferType = null;
			$wrapUpAnyway  = false;
			switch ($context) {
				//Inside normal code.
				case 'code': {
					//Comments
					if ($c == '/' && $nc == '/') {
						$context = 'comment.line';
						break;
					}
					if ($c == '/' && $nc == '*') {
						$context = 'comment.block';
						break;
					}
					
					//Strings
					if ($c == '"') {
						$context = 'string';
						break;
					}
					
					//Symbols
					$isSymbol = in_array(ord($c), Language::$symbols);
					if ($isSymbol) {
						//Mark the buffer as being a symbol and to be wrapped up immediately,
						//since we want symbols to be separate tokens, not clumped together.
						$newBufferType = 'symbol';
						$wrapUpAnyway = true;
						
						//Check whether the symbol would make a valid symbol combination with
						//what's already in the buffer, in which case we leave the symbols
						//clumped together.
						$lookaheadBuffer = $buffer . $c;
						if (in_array($lookaheadBuffer, Language::$symbolCombinations)) {
							$wrapUpAnyway = false;
						}
						break;
					}
					
					//Numbers
					if (ord($c) >= 0x30 && ord($c) <= 0x39) {
						$newBufferType = 'numeric';
						break;
					}
					
					//Identifiers (anything but the above and whitespaces)
					if ($c != null && strchr(" \t\n\r", $c) === false) {
						$newBufferType = 'identifier';
						break;
					}
				} break;
				
				//Inside comments.
				case 'comment.line': {
					if ($c == "\n")
						$context = 'code';
				} break;
				case 'comment.block': {
					if ($pc == '*' && $c == '/')
						$context = 'code';
				} break;
				
				//Inside strings.
				case 'string': {
					//Check whether we've reached the end of the string in which case we switch
					//back to code context. We leave new buffer kind invalid so the token gets
					//wrapped up.
					if (($ppc == '\\' || $pc != '\\') && $c == '"') {
						$context = 'code';
					}
					
					//Otherwise we mark this character as part of a string.
					else {
						$newBufferType = 'string';
					}
				} break;
			}
			
			//Numeric tokens are allowed inside an identifier.
			if ($bufferType == 'identifier' && $newBufferType == 'numeric') {
				$newBufferType = $bufferType;
			}
			
			//Numeric buffers suck up all identifiers and some symbol tokens.
			if ($bufferType == 'numeric') {
				if ($newBufferType == 'identifier' || $c == '.' ||
					(strchr("eE", $pc) !== false && strchr("+-", $c) !== false)) {
					$newBufferType = $bufferType;
					$wrapUpAnyway = false;
				}
			}
			
			//If the current and new buffer types disagree, we have to wrap up the current
			//buffer.
			if (($bufferType != $newBufferType || $wrapUpAnyway) && $buffer != '') {
				
				//Buffer type alterations based on content.
				if (in_array($buffer, Language::$identifierSymbols)) {
					$bufferType = 'identifier';
				}
				if (in_array($buffer, Language::$keywords)) {
					$bufferType = 'keyword';
				}
				
				//Create the default token and add it to the flat token list.
				$dt = new Token;
				$dt->type  = $bufferType;
				$dt->range = clone $bufferRange;
				$dt->text  = $buffer;
				$this->flatTokens[] = $dt;
				
				//If the current token is a symbol and it closes the topmost group on the stack,
				//remove the group and ignore the symbol.
				if ($bufferType == 'symbol' && strchr(")}]", $buffer) !== false) {
					if (count($groupStack) > 0) {
						$style = $groupStack[0]->text;
						if ($style[1] != $buffer) {
							$this->issues[] = "$bufferRange: symbol $buffer doesn't close previous group $style properly";
						} else {
							$groupStack[0]->range->combine($bufferRange);
							array_shift($groupStack);
						}
					} else {
						$this->issues[] = "$bufferRange: symbol $buffer does not have a corresponding opening symbol";
					}
				}
				
				//Otherwise we create a new node for this token.
				else {
					//If this is a bracket symbol, create a new group token.
					if ($bufferType == 'symbol' && strchr("({[", $buffer) !== false) {
						$styles = array('(' => '()', '{' => '{}', '[' => '[]');
						$dt = new TokenGroup;
						$dt->type  = 'group';
						$dt->range = clone $bufferRange;
						$dt->text = $styles[$buffer];
					}
					
					//Add the node to the topmost group.
					if (count($groupStack) > 0) {
						$groupStack[0]->tokens[] = $dt;
					} else {
						$this->tokens[] = $dt;
					}
					
					//If the new node is a group, add it to the top of the stack.
					if ($dt->is('group')) {
						array_unshift($groupStack, $dt);
					}
				}
				
				//Clear the buffer and buffer type.
				$buffer     = '';
				$bufferType = null;
			}
			
			//If the current token is valid, append it to the buffer.
			if ($newBufferType != null) {
				if ($bufferType == null) {
					$bufferRange->start = clone $loc;
				}
				$bufferType = $newBufferType;
				$buffer .= $c;
			}
			
			//Increase the location column counter.
			if ($c == "\n") {
				$loc->line++;
				$loc->column = 0;
			} else {
				$loc->column++;
			}
			$loc->offset++;
			
			//Adjust the buffer range's end.
			$bufferRange->end = clone $loc;
		}
		
		foreach ($this->issues as $i) {
			echo $i."\n";
		}
	}
}
